b_cfn_elasticsearch_cloner/resource.py

```diff
@@ -0,0 +1,211 @@
+                                    "es:ESHttpPost",
+                                    "es:ESHttpPut",
+                                ],
+                                resources=["*"],
+                                effect=Effect.ALLOW,
+                            )
+                        ]
+                    ),
+                    "DynamodbPolicy": PolicyDocument(
+                        statements=[
+                            PolicyStatement(
+                                actions=["dynamodb:*"],
+                                resources=["*"],
+                                effect=Effect.ALLOW,
+                            )
+                        ]
+                    ),
+                },
+                description="Role for DynamoDB Initial Cloner Function",
+            ),
+        )
+
+        if kms_key:
+            initial_cloner_function.add_to_role_policy(
+                PolicyStatement(
+                    actions=["kms:Decrypt"],
+                    resources=[kms_key.key_arn],
+                    effect=Effect.ALLOW,
+                ),
+            )
+
+        initial_cloner = CustomResource(
+            scope=self,
+            id="InitialCloner",
+            service_token=initial_cloner_function.function_arn,
+            removal_policy=RemovalPolicy.DESTROY,
+            properties={
+                "DynamodbTableName": dynamodb_table.table_name,
+                "ElasticsearchIndexName": elasticsearch_index.index_name,
+                "ElasticsearchEndpoint": elasticsearch_index.elasticsearch_domain.domain_endpoint,
+            },
+            resource_type="Custom::ElasticsearchInitialCloner",
+        )
+
+        primary_key_field = initial_cloner.get_att_string("PrimaryKeyField")
+
+        dynamodb_stream_arn = dynamodb_table.table_stream_arn
+        dynamodb_event_source = DynamoEventSource(
+            table=dynamodb_table,
+            starting_position=StartingPosition.LATEST,
+            enabled=True,
+            max_batching_window=Duration.seconds(10),
+            bisect_batch_on_error=True,
+            parallelization_factor=2,
+            batch_size=1000,
+            retry_attempts=10,
+        )
+
+        cloner_function = Function(
+            scope=self,
+            id="ClonerFunction",
+            code=Code.from_asset(cloner_root),
+            handler="index.handler",
+            runtime=Runtime.PYTHON_3_8,
+            environment={
+                "ES_INDEX_NAME": elasticsearch_index.index_name,
+                "ES_DOMAIN_ENDPOINT": elasticsearch_index.elasticsearch_domain.domain_endpoint,
+                "PRIMARY_KEY_FIELD": primary_key_field,
+            },
+            events=[dynamodb_event_source],
+            layers=[ElasticsearchLayer],
+            log_retention=RetentionDays.ONE_MONTH,
+            memory_size=128,
+            role=Role(
+                scope=self,
+                id="ClonerFunctionRole",
```

Comment:
Do not hardcode ids. Two same resources would result in a CDK error. Instead of taking an "id" in the constructor, take a prefix and add that prefix to all of your resource names and ids.